### ISSUE TITLE
Convert C10_Complex / CppComplex to Nim Complex[T] on high level API

### ISSUE THREAD

**Issue by Clonkk** [SciNim/flambeau#24]
Latest examples is using ``data_ptr`` : it return an ``ptr UncheckedArray[C10_Complex[float]]``. 

For ease of use, it is better to convert it to ``ptr UncheckedArray[Complex]``.

``item`` procs, various FFTs procs already have special implementation for Complex; a code review should be done on the API to make sure Complex is handled properly on the idiomatic Nim API

**Comment by HugoGranstrom** [comment]
I'm a bit torn on this. `data_ptr` should be a pointer to the underlying data, right? If we convert it to `Complex` it is an entirely new piece of data it is pointing to. I think the conversion in this case when the user is already handling low-level stuff (pointers), should be manual. Otherwise, the user is forced to do unnecessary copies when it may not be necessary. So a `proc UncheckedComplex(a: ptr UncheckedArray[C10_Complex[float]]): ptr UncheckedArray[Complex]` would be better in my opinion (but a better name).

**Comment by Clonkk** [comment]
Not necessarily; since ``sizeof(C10_Complex[float32]) == sizeof(Complex[float32])`` the following code works : 
```nim
import flambeau
import complex
var A = initC10_Complex[float32](3.0, 1.0)
var B = Complex[float32](re: 3.0.float32, im: 1.0.float32)
echo A
echo B
doAssert sizeof(A) == sizeof(B)

var pA = A.addr
block:
  var ua = cast[ptr UncheckedArray[float32]](pA)
  echo ua[0]
  echo ua[1]
  doAssert ua[0] == B.re
  doAssert ua[1] == B.im

block:
  var ua = cast[ptr UncheckedArray[Complex[float32]]](pA)
  echo ua[0]
  doAssert ua[0] == B
``` 
So specializing ``data_ptr`` for Complex type  and casting should work

For further Complex handling, we could also specialize the ``Tensor`` type for complex : 
```nim
type
  Tensor[Complex32] = object
    raw: RawTensor # RawTensor that holds c10 data
```
Then we'd probably need to expose the construction via  https://pytorch.org/docs/stable/generated/torch.complex.html for instance.

**Comment by HugoGranstrom** [comment]
Oh so they are byte-compatible? 😯 So we could cast a C10_Complex to a Complex? Then I agree that a cast is appropriate 👍

**Comment by Clonkk** [comment]
> we could cast a C10_Complex to a Complex?

Just to be precise (this is probably what you had in mind), we can cast a ``ptr C10_Complex`` to a ``ptr Complex``.

**Comment by HugoGranstrom** [comment]
Nah, I meant what I wrote. If we cast `ptr C10_Complex` to `ptr Complex` we expect to be able to dereference it. And then it should behave as a `Complex`. In other words, must `C10_Complex` be cast-able to `Complex`.

**Comment by Clonkk** [comment]
Casting non-pointer type is incorrect most of the time in C++ (with a few exceptions like casting between integer types) . 

This is incorrect (and does not compile) : 
```nim
let c1 = initC10_Complex[float32](1.0, 2.0)
let c2 : Complex32 = cast[Complex32](c)
```

This is the correct way to convert between byte-compatible type :
```nim
var c1 = initC10_Complex[float32](1.0, 2.0)
var c2 : Complex32 = cast[ptr Complex32](addr(c))[]
```

**Comment by HugoGranstrom** [comment]
Hmm ok, didn't know C++ was so fussy about type safety for once 🤣 
The important part is that there is a way to make it work at least 😄
